osu.Game/Skinning/SkinnableSprite.cs

```diff
@@ -1,46 +1,102 @@
+using osu.Game.Configuration;
+using osu.Game.Overlays.Settings;
+using osuTK;
 
 namespace osu.Game.Skinning
 {
     /// <summary>
     /// A skinnable element which uses a stable sprite and can therefore share implementation logic.
     /// </summary>
-    public class SkinnableSprite : SkinnableDrawable
+    public class SkinnableSprite : SkinnableDrawable, ISkinnableDrawable
     {
         protected override bool ApplySizeRestrictionsToDefault => true;
 
         [Resolved]
         private TextureStore textures { get; set; }
 
+        [SettingSource("Sprite name", "The filename of the sprite", SettingControlType = typeof(SpriteSelectorControl))]
+        public Bindable<string> SpriteName { get; } = new Bindable<string>(string.Empty);
+
+        [Resolved]
+        private ISkinSource source { get; set; }
+
         public SkinnableSprite(string textureName, ConfineMode confineMode = ConfineMode.NoScaling)
             : base(new SpriteComponent(textureName), confineMode)
         {
+            SpriteName.Value = textureName;
+        }
+
+        public SkinnableSprite()
+            : base(new SpriteComponent(string.Empty), ConfineMode.NoScaling)
+        {
+            RelativeSizeAxes = Axes.None;
+            AutoSizeAxes = Axes.Both;
+
+            SpriteName.BindValueChanged(name =>
+            {
+                ((SpriteComponent)Component).LookupName = name.NewValue ?? string.Empty;
+                if (IsLoaded)
+                    SkinChanged(CurrentSkin);
+            });
         }
 
         protected override Drawable CreateDefault(ISkinComponent component)
         {
             var texture = textures.Get(component.LookupName);
 
             if (texture == null)
-                return null;
+            {
+                return new SpriteIcon
+                {
+                    Size = new Vector2(100),
+                    Icon = FontAwesome.Solid.QuestionCircle
+                };
+            }
 
             return new Sprite { Texture = texture };
         }
 
+        public bool UsesFixedAnchor { get; set; }
+
         private class SpriteComponent : ISkinComponent
         {
+            public string LookupName { get; set; }
+
             public SpriteComponent(string textureName)
             {
                 LookupName = textureName;
             }
+        }
 
-            public string LookupName { get; }
+        public class SpriteSelectorControl : SettingsDropdown<string>
+        {
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                // Round-about way of getting the user's skin to find available resources.
+                // In the future we'll probably want to allow access to resources from the fallbacks, or potentially other skins
+                // but that requires further thought.
+                var highestPrioritySkin = ((SkinnableSprite)SettingSourceObject).source.AllSources.First() as Skin;
```

Comment:
This line is my biggest hang-up here for several reasons:

* Pretty ad-hoc access to a drawable (the skinnable sprite) that could be in any state, including detached from hierarchy/disposed/whichever
* Assumption that the first source is the user skin, which I'm not even sure of, since I would have guessed it would be the beatmap skin - unless this is the case in skin editor (in skin editor only?)

Dunno how to fix it immediately, because I'd probably have to sit down and try to rewrite all of this to see where I end up at, but maybe you'll have some ideas on how to make this less magic?

Reply:
I don't disagree with anything mentioned here, but I was hoping to get things working with magic because there are a lot more pieces to hook up before I'm ready to refactor things.

For instance I have similar magic/hacky logic in another upcoming PR I am going to make, but more focused on getting things in place so I can get the UX into the hands of users (and myself) to see what is required in the bigger picture.

There's definitely going to need to be some large refactoring to, for instance, support user overrides of certain skin settings; changes to potentially avoid doing full skin reloads on texture changes; changes to `RealmFileUsage` structure (they can't be `EmbeddedObject` for instance) to allow querying all sprites of the same lookup name across all skins (to allow users to "mix" skins).

Is also why I left a large inline comment for this as a todo-ish area.

Reply:
Upon testing, it turns out that one of the above assumptions (re: the skin source ordering) is indeed a bit flimsy. If you open the skin editor during gameplay and try to switch the texture on a sprite by the dropdown on the right, it will be empty, because the skins are being wrapped by transformers:

![2022-04-04-224153_872x308_scrot](https://user-images.githubusercontent.com/20418176/161630280-39788436-74e0-4ffa-a6ad-62f41ed7fcb5.png)

If you're looking for a "make work" solution, here's one:

```diff
diff --git a/osu.Game/Skinning/LegacySkinTransformer.cs b/osu.Game/Skinning/LegacySkinTransformer.cs
index 97084f34e0..9481fc7182 100644
--- a/osu.Game/Skinning/LegacySkinTransformer.cs
+++ b/osu.Game/Skinning/LegacySkinTransformer.cs
@@ -23,7 +23,7 @@ public abstract class LegacySkinTransformer : ISkin
         /// The <see cref="ISkin"/> which is being transformed.
         /// </summary>
         [NotNull]
-        protected ISkin Skin { get; }
+        protected internal ISkin Skin { get; }
 
         protected LegacySkinTransformer([NotNull] ISkin skin)
         {
diff --git a/osu.Game/Skinning/SkinnableSprite.cs b/osu.Game/Skinning/SkinnableSprite.cs
index c5f110f908..4b4d7fe2c6 100644
--- a/osu.Game/Skinning/SkinnableSprite.cs
+++ b/osu.Game/Skinning/SkinnableSprite.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
+using System.Collections.Generic;
 using System.Linq;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
@@ -83,7 +84,7 @@ protected override void LoadComplete()
                 // Round-about way of getting the user's skin to find available resources.
                 // In the future we'll probably want to allow access to resources from the fallbacks, or potentially other skins
                 // but that requires further thought.
-                var highestPrioritySkin = ((SkinnableSprite)SettingSourceObject).source.AllSources.First() as Skin;
+                var highestPrioritySkin = getHighestPriorityUserSkin(((SkinnableSprite)SettingSourceObject).source.AllSources) as Skin;
 
                 string[] availableFiles = highestPrioritySkin?.SkinInfo.PerformRead(s => s.Files
                                                                                           .Where(f => f.Filename.EndsWith(".png", StringComparison.Ordinal)
@@ -92,6 +93,26 @@ protected override void LoadComplete()
 
                 if (availableFiles?.Length > 0)
                     Items = availableFiles;
+
+                static ISkin getHighestPriorityUserSkin(IEnumerable<ISkin> skins)
+                {
+                    foreach (var skin in skins)
+                    {
+                        if (skin is LegacySkinTransformer transformer && isUserSkin(transformer.Skin))
+                            return transformer.Skin;
+
+                        if (isUserSkin(skin))
+                            return skin;
+                    }
+
+                    return null;
+                }
+
+                // Temporarily used to exclude undesirable ISkin implementations
+                static bool isUserSkin(ISkin skin)
+                    => skin.GetType() == typeof(DefaultSkin)
+                       || skin.GetType() == typeof(DefaultLegacySkin)
+                       || skin.GetType() == typeof(LegacySkin);
             }
         }
 
```

Reply:
I've pushed the above code verbatim.